py/src/nupic/tests/encoders/date_test.py

```diff
@@ -0,0 +1,232 @@
+    # 14:55 is minute 14*60 + 55 = 895; 895/48 = cetral bit= bit 18.6 ->18 #FIXME we should round this, so becomes bit19
+    # We activate our 5(=w)bits around the central bit -> thus [18-(5-1)/2, 18+(5-1)/2] = [16, 20]
+    # Encoder should be 30 bits total (30 * 48 minutes = 24 hours)
+    timeOfDayExpected = (
+      [0,0,0,0,0,0,0,0,0,0,
+       0,0,0,0,0,0,1,1,1,1,
+       1,0,0,0,0,0,0,0,0,0])
+
+    expected = timeOfDayExpected
+    self.assertEqual(bits.size, 30)
+    self.assertEqual(expected, bits.dense.tolist())
+
+
+  def testMissingValues(self):
+    """ Missing values. """
+    e = DateEncoder(timeOfDay=5)
+    mvOutput = e.encode(None)
+    self.assertEqual(sum(mvOutput.sparse), 0)
+    mvOutput = e.encode(float('nan'))
+    self.assertEqual(sum(mvOutput.sparse), 0)
+
+
+  def testHoliday(self):
+    """ Look at holiday more carefully because of the smooth transition. """
+    e = DateEncoder(holiday=5)
+    holiday    = [0,0,0,0,0,1,1,1,1,1]
+    notholiday = [1,1,1,1,1,0,0,0,0,0]
+    holiday2   = [0,0,0,1,1,1,1,1,0,0]
+
+    d = datetime.datetime(2010, 12, 25, 4, 55) #Christmas day 25th Dec, a default holiday
+    assert(all( e.encode(d).dense == holiday ))
+
+    d = datetime.datetime(2008, 12, 27, 4, 55) #12/27 is not a holiday
+    assert(all( e.encode(d).dense == notholiday ))
+
+    d = datetime.datetime(1999, 12, 26, 8, 00) #day after holiday, approaching
+    assert(all( e.encode(d).dense == holiday2 ))
+
+    d = datetime.datetime(2011, 12, 24, 16, 00) #day before holiday, approaching
+    assert(all( e.encode(d).dense == holiday2 ))
+
+
+  def testHolidayMultiple(self):
+    """ Look at holiday more carefully because of the smooth transition. """
+    e = DateEncoder(holiday=5, holidays=[(12, 25), (2018, 4, 1), (2017, 4, 16)])
+    holiday    = [0, 0, 0, 0, 0, 1, 1, 1, 1, 1]
+    notholiday = [1, 1, 1, 1, 1, 0, 0, 0, 0, 0]
+
+    d = datetime.datetime(2011, 12, 25, 4, 55)
+    assert(all( e.encode(d).dense == holiday ))
+
+    d = datetime.datetime(2007, 12, 2, 4, 55)
+    assert(all( e.encode(d).dense == notholiday ))
+
+    d = datetime.datetime(2018, 4, 1, 16, 10)
+    assert(all( e.encode(d).dense == holiday ))
+
+    d = datetime.datetime(2017, 4, 16, 16, 10)
+    assert(all( e.encode(d).dense == holiday ))
+
+
+  def testWeekend(self):
+    """ Test weekend encoder. """
+    e  = DateEncoder(customDays=(21, ["sat", "sun", "fri"]))
+    e2 = DateEncoder(weekend=21)
+
+    d = datetime.datetime(1988, 5, 29, 20, 00)
+    print(d)
+    self.assertEqual( e.encode(d), e2.encode(d) )
+
+    for _ in range(300):
+      d = d+datetime.timedelta(days=1)
+      self.assertEqual( e.encode(d), e2.encode(d) )
+
+
+  def testYearsDiffer(self):
+    """ Creating date encoder instance. """
+    enc = DateEncoder(season=1, dayOfWeek=1, weekend=1) #all info for recognizing days 
+    # 1.1. 2007 & 2018 was Monday, can you recognize the days?
+    first2007 = datetime.datetime(2007, 1, 1) #FIXME enc fails to encode this? 
```

Comment:
for some reason, encoder fails when dealing with these dates (fails to even encode)

Reply:
@shahardror yes, this is the issue you've discovered in #578